File: leetcode_2021/leetcode_cpp/16.最近的三数之和.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int i = 0;
        int res = nums[0] + nums[1] + nums[2];
        int min_abs = abs(res - target);
        for (auto iter = nums.begin(); iter <= nums.end() - 3;) {
            auto left = iter + 1;
            auto right = nums.end() - 1;
            while (left < right) {
                int val = *iter + *left + *right;
                if (val == target) {
                    return val;
                } else if (val < target) {
                    if (abs(val - target) < min_abs) {
                        min_abs = abs(val - target);
                        res = val;
                    }
                    left += 1;
                } else {
                    if (abs(val - target) < min_abs) {
                        min_abs = abs(val - target);
                        res = val;
                    }
                    right -= 1;
                }
            }
            iter += 1;
        }
        return res;
    }
};

File: leetcode_2021/leetcode_python/16.最近的三数之和.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        i = 0
        res = sum(nums[:3])
        min_abs = abs(res - target)
        while i <= len(nums) - 3:
            left, right = i + 1, len(nums) - 1
            while left < right:
                val = nums[i] + nums[left] + nums[right]
                if val == target:
                    return val
                elif val < target:
                    if abs(val - target) < min_abs:
                        min_abs = min(min_abs, abs(val - target))
                        res = val
                    left += 1
                else:
                    if abs(val - target) < min_abs:
                        min_abs = min(min_abs, abs(val - target))
                        res = val
                    right -= 1
            i += 1
        return res